src/axom/core/Array.hpp

```diff
@@ -1484,6 +1509,37 @@ inline T* Array<T, DIM, SPACE>::reserveForInsert(IndexType n, IndexType pos)
   return m_data + pos;
 }
 
+//------------------------------------------------------------------------------
+template <typename T, int DIM, MemorySpace SPACE>
+AXOM_DEVICE inline IndexType Array<T, DIM, SPACE>::reserveForDeviceInsert(IndexType n)
+{
+#ifndef AXOM_DEVICE_CODE
+  // Host path: should never be called.
+  AXOM_UNUSED_VAR(n);
+  assert(false);
+  return {};
+#else
+  // Device path: supports insertion while m_num_elements < m_capacity
+  // Does not support insertions which require reallocating the underlying
+  // buffer.
+  IndexType new_pos = RAJA::atomicAdd<RAJA::auto_atomic>(&m_num_elements, n);
+  if(new_pos >= m_capacity)
+  {
+  #ifdef AXOM_DEBUG
+    printf(
+      "Array::reserveForInsert: size() exceeded capacity() when inserting "
+      "on the device.\n");
```

Comment:
Style note only: I'm a little surprised at the use of `printf()`, especially in device code, but I'm not sure what else there is.